### ISSUE TITLE
Proposal: Reduce the components for simple architecture

### ISSUE THREAD

**Issue by supereagle** [caicloud/cyclone#285]
## Current components
* **Clair + Postgres**: for image security, not needed any more
* **Kafka + Zookeeper**: for record real-time log
* **ETCD**: for the queue of jobs
* **MongoDB**: for data persistence

## Target
Only keep the MongoDB for data persistence, and remove all other components.

## Solution
* **~ETCD~**: Use MongoDB for the queue of jobs, please refer to [db.collection.findAndModify()](https://docs.mongodb.com/v3.4/reference/method/db.collection.findAndModify/). Investigation in #286 
* **~Kafka + Zookeeper~**: The record log is stored in file, if want to get the real-time log, watch the change of the file content. Investigation in #288 
* **~Clair + Postgres~**: Can be directly removed as image security is not needed any more.

### Architecture

![image](https://user-images.githubusercontent.com/3390333/34285920-460fd06e-e718-11e7-9d58-32e192d84ac6.png)

**Comment by zoumo** [comment]
How to use MongoDB as the queue of jobs by db.collection.findAndModify()?

**Comment by superxi911** [comment]
Polling?
> How to use MongoDB as the queue of jobs by db.collection.findAndModify()?

**Comment by supereagle** [comment]
@zoumo @superxi911   When the job comes, the server will try to handle it immediately. If the job can not be handled for some reasons like cluster is busy, it will be added into MongoDB. 

A goroutine in the server will periodically get the pending jobs from MongoDB and handle it. If the job can not be handled again, it will be added back to MongoDB with lower priority.

**Comment by superxi911** [comment]
If some jobs are being handled，the server and works crash with some reason. Then, the server restart. What will happen?

**Comment by supereagle** [comment]
What the server needs to do while the job is running:
* Receive the real-time log
* Receive the status of the job

What the worker needs to do while the job is running:
* Run the job 
* Send the real-time log to server
* Report job status to server

There 3 exceptional cases while the job is running:
* Server is health but worker crashes: Server will set the job as failed after no response from worker for a long time.
* Server crashes but worker is health: Worker will not be affected. In order to avoid the missing of log and status from worker, there will be several server instances.
* Both server and worker crash: Equals to the 1st case when there are several server instances.

**Comment by LiangXugang** [comment]
# refactor of event manager 
**now**
the event manager with `etcd watcher` and `queue in memory`

**refactor target**

- remove etcd 

- persist the queue

- concurrency security 

- stateless event manager for horizontal scale-up

**mongoMQ**

based on mongoDB `findAndModify` method

- atomic operation

- find + update at the sametime

- return the original data

doc: https://docs.mongodb.com/v3.4/reference/method/db.collection.findAndModify/

_data type_

the queue object as this:

```
type Queue interface {
    In(*Event)                      // enqueue event
    Out() (*Massage, error) // dequeue  event
    ReIn(*Massage)            // re-enqueue event when busy
    Remove(*Massage)      // event uesd and remove the record
}
```

the massage in mongoMQ as this:

```
type Massage struct {
    ID       string   // _id
    Event *Event // event data
    State  int        // inQueue or outQueue
    Time   int64   // timestamp, priority of the event. UnixTime when created for the first time
    Retry  int       // retry times, delay duration denpending on this
}
```

_architecture_
- poll massage from mongo, get the event with highest priority( the most recent timestamp)

- dequeue and use massage only once (set `Massage.State` to `outQ`), if massage used, remove the massage in mongoDB( or set it to just recorded?).

- re-enqueue the event when busy, reduce the priority of the event(update the massage in mongoDB -- `Massage.Retry ++`, delay` Massage.Time` based on `Massage.Retry` and set `Massage.State` to `inQ`)

_mongoMQ verification demo_
https://github.com/LiangXugang/demo/tree/master/mongoMQ


**event manager with mongoDB MQ**

- handle events received (or just receive events and skip to 2nd step)

- send the event to mongoMQ (when busy or usual)

- poll event from mongoMQ and make event work

- event working failed with busy cloud, resend to mongoMQ

**Comment by LiangXugang** [comment]
@supereagle @zoumo @superxi911

**Comment by supereagle** [comment]
Why the [mongoMQ verification demo](https://github.com/LiangXugang/demo/tree/master/mongoMQ) does not use 'findAndModify()'?

> based on mongoDB findAndModify method
> * atomic operation
> * find + update at the sametime
> * return the original data

**Comment by LiangXugang** [comment]
the method `findAndModify()` is in mongoDSL
`Apply()` is the corresponding method in `"gopkg.in/mgo.v2"`

_references:_
[How to realize `Apply()`](http://bazaar.launchpad.net/+branch/mgo/v2/view/head:/session.go#L3172)
[how-can-i-findandmodify-one-record-in-mongodb-with-golang](https://stackoverflow.com/questions/31937292/how-can-i-findandmodify-one-record-in-mongodb-with-golang)
[mgo-doc](https://godoc.org/labix.org/v2/mgo)

**Comment by zoumo** [comment]
@LiangXugang @supereagle 
After getting an event from mongodb, and setting Massage.State to outQ, the server crashs. worker is not started by server. 
How to deal with this issue? Maybe need a timeout

**Comment by LiangXugang** [comment]
**Architecture**

![mongomq 1](https://user-images.githubusercontent.com/33119412/34472706-9794fce2-efa2-11e7-9018-a404d45b0e43.jpg)

**Comment by supereagle** [comment]
Have implemented.
/close

**Comment by supereagle** [comment]
Need docs for this architecture.
/reopen
